fairmq/shmem/Manager.h

```diff
@@ -665,9 +693,14 @@ class Manager
     Uint16SegmentInfoHashMap* fShmSegments;
     Uint16RegionInfoHashMap* fShmRegions;
     std::unordered_map<uint16_t, std::unique_ptr<Region>> fRegions;
+    // make sure this is alone in the cache line: mostly read
+    alignas(128) inline static std::atomic<unsigned long> fRegionsGen = 0ul;
```

Comment:
Does the `alignas` really do what the comment says? It aligns the long, but since there is no padding after, something can land after it in the cache line?

Reply:
No, you're right. Padding must be manually added.

Reply:
Thanks to @mattkretz for noticing.

Reply:
Late to the party. But I wanted to point out the interesting difference of declaring a member with overalignment and declaring a type with overalignment. The latter also affects the `sizeof` of the type (adding padding), so that a C-array of the type has the correct alignment for all entries. The former also makes the sizeof of the enclosing type a multiple of the largest alignment but *not* the overaligned member.
```c++
struct alignas(64) A {};
struct B {
    alignas(64) int a;
};

// 64 == sizeof(A);
// 64 == sizeof(B);
//  4 == sizeof(B::a);
```

Reply:
Oh, and overalignment can potentially waste a lot of memory. If possible (basically only construction order and ABI compatibility concerns), reorder the members from largest alignment to smallest alignment.

Reply:
We have two use cases for over alignment:  
1 nasty counters that keep bouncing the cache line when changed from multiple threads; isolate from the rest of data
2 stable and often used values that could happily live in many caches in the shared state: prevent neighboring variables to invalidate the cache line.

Memory wasting is not usually the worry. Cache line should be mostly 64B, but I remember Intel prefetchers always working with aligned 128B of data (2 CL).

Reply:
Oh, I missed that `fRegionsGen` is `static`. I.e. it doesn't affect the `sizeof` (and padding) of this structure at all. But AFAIK, it still doesn't ensure that other global objects don't go into the same cacheline (behind it).

Reply:
Yes, strange rules, but prob for a good reason. Unfortunately, `alignas` does not work with typedefs to change the `sizeof` of existing types. It has to be a struct or union...

Reply:
I don't find it strange, after all you're just asking for a certain alignment of the variable, not for any padding. The compiler might need to add padding before the object to achieve the alignment, but that's it. If you require padding after the atomic you must use a structure. E.g.:
```c++
struct A {
    inline static struct alignas(128) {
        int x;
    } x = {0};
};
```
https://godbolt.org/z/nnb9PPoPo